x-pack/plugins/apm/public/context/UrlParamsContext/index.tsx

```diff
@@ -4,74 +4,40 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import React, { createContext, useReducer, useEffect } from 'react';
-import { Location } from 'history';
-import { useLocation } from '../../hooks/useLocation';
+import React, { createContext } from 'react';
+import { History } from 'history';
 import { IUrlParams } from './types';
-import { LOCATION_UPDATE, TIME_RANGE_REFRESH } from './constants';
-import { getParsedDate } from './helpers';
 import { resolveUrlParams } from './resolveUrlParams';
+import { fromQuery, toQuery } from '../../components/shared/Links/url_helpers';
 
 interface TimeRange {
   rangeFrom: string;
   rangeTo: string;
 }
 
-interface LocationAction {
-  type: typeof LOCATION_UPDATE;
-  location: Location;
-}
-
-interface TimeRangeRefreshAction {
-  type: typeof TIME_RANGE_REFRESH;
-  time: TimeRange;
-}
-
 const defaultRefresh = (time: TimeRange) => {};
 
-export function urlParamsReducer(
-  state: IUrlParams = {},
-  action: LocationAction | TimeRangeRefreshAction
-): IUrlParams {
-  switch (action.type) {
-    case LOCATION_UPDATE: {
-      return resolveUrlParams(action.location, state);
-    }
-
-    case TIME_RANGE_REFRESH:
-      return {
-        ...state,
-        start: getParsedDate(action.time.rangeFrom),
-        end: getParsedDate(action.time.rangeTo)
-      };
-
-    default:
-      return state;
-  }
-}
-
 const UrlParamsContext = createContext({
   urlParams: {} as IUrlParams,
   refreshTimeRange: defaultRefresh
 });
 
-const UrlParamsProvider: React.FC<{}> = ({ children }) => {
-  const location = useLocation();
-  const [urlParams, dispatch] = useReducer(
-    urlParamsReducer,
-    resolveUrlParams(location, {})
-  );
-
-  function refreshTimeRange(time: TimeRange) {
-    dispatch({ type: TIME_RANGE_REFRESH, time });
-  }
-
-  useEffect(
-    () => {
-      dispatch({ type: LOCATION_UPDATE, location });
-    },
-    [location]
-  );
+const UrlParamsProvider: React.FC<{ history: History }> = ({
+  history,
+  children
+}) => {
+  const urlParams = resolveUrlParams(history.location, {});
+
+  const refreshTimeRange = (time: TimeRange) => {
+    history.replace({
+      ...history.location,
+      search: fromQuery({
+        ...toQuery(history.location.search),
+        start: time.rangeFrom,
+        end: time.rangeTo
+      })
+    });
+  };
```

Comment:
I think this will break the "refresh" button. If the user has a relative time range `rangeFrom` and `rangeTo` will not change and so the url will not change.

Reply:
Btw. `urlParams ` are derived from the url so should only ever read from it - not change it.

Reply:
I tested that, and it seems to still work. It basically just refreshes the current URL, triggering a re-render via the `Router` component, and then the start/end values are re-calculated again.

Reply:
I think this could just be an optimization away from breaking. It might be because whenever react-router updates, all context providers are currently not using memo as suggested in https://github.com/elastic/kibana/pull/35939. 

Eg. if the relative date hasn't changed we don't re-calculate the absolute date:
https://github.com/elastic/kibana/blob/3efd42c63b67f6d8ba11a79abd6e087d457279ba/x-pack/plugins/apm/public/context/UrlParamsContext/helpers.ts#L21-L24